### ISSUE TITLE
CodeAction.Create should not have an equivalence key

### ISSUE THREAD

**Issue by jaredpar** [dotnet/roslyn#5975]
Consider this overload of CodeAction.Create

``` csharp
public static CodeAction Create(string title, IEnumerable<CodeAction> nestedActions, string equivalenceKey = null)
```

During API review it seemed that there was no reason for an equivalenceKey here.  The default is to use the first `CodeAction` so shouldn't it always be the equivalence key.

**Comment by bkoelman** [comment]
Does removing equivalenceKey here make my usage (https://github.com/dotnet/roslyn/issues/4078) impossible?

**Comment by Pilchie** [comment]
No, this is specifically about the overload that has the nested actions for displaying a submenu.

**Comment by bkoelman** [comment]
@Pilchie Ah yes, I see now. Sorry to interrupt.

**Comment by CyrusNajmabadi** [comment]
I think the equivalence key should remain.  It still makes sense that multiple actions might try to present two of these with teh same equivalence key, and we'd then want to dedupe htem.  I don't believe it's always the case that that means their first actual CodeAction's equivalence key would then be the same.

It is true that for the *single* CodeAction (generate type) that uses this new facility we do just pass along the equivalence key for the first action.  But i'm not at all convinced that that is universal, or that we should take away the equivalence key here.  I would argue this should stay as is.

**Comment by Pilchie** [comment]
That's not what equivalence key is used for.  It's used to determine which fixes "line up" when invoking fix-all.  In this case there is nothing to "apply" for this action in a fix-all, so it can't be equivalent to something else.

**Comment by CyrusNajmabadi** [comment]
Ah, I did not realize that.  Then definitely this should not be there... But, if that's the case, it's not clear that we'd want to pull in the first actions' equivalence key either.  Sounds like we should just have the concept that a code action may not have an equivalence key.  And we only try to "line up" ones that do?

**Comment by mavasani** [comment]
@Pilchie Equivalence key is currently used for 2 purposes:

1. The FixAll case you mentioned above.

2. De-duping the suggested actions [here](http://source.roslyn.io/#Microsoft.CodeAnalysis.EditorFeatures/Implementation/Suggestions/SuggestedAction.cs,299).

I have added a relevant comment [here](https://github.com/dotnet/roslyn/pull/6238#discussion_r42769972).

**Comment by CyrusNajmabadi** [comment]
Hrmm..  if it is for deduping... then i'd think we'd still want this equivalence key.

I'm not sure how this will affect FixAll though...

**Comment by mavasani** [comment]
Yes, we still want non-empty equivalence key for top level actions, and default should be based on nested actions.

Fix all really doesn't make sense for a top level action as it is a collection of non-identical nested actions and we wouldn't know which nested action to apply. We should not even compute a fix all suggested action for a top level action, i.e. bail out [here](http://source.roslyn.io/#Microsoft.CodeAnalysis.EditorFeatures/Implementation/Suggestions/CodeFixSuggestedAction.cs,54) if `action` has nested actions.

**Comment by CyrusNajmabadi** [comment]
Note: from VS's perspective, a top level item can be an action itself.  Though we currently don't expose any way to accomplish that with our CodeAction API.  

i.e. within VS you could hav a top level action which you could click on.  And that top level action could also have nested actions as well.  

If we expose that through our API, we'd want to have a top level action which also had a collection of children.  And that top level action would need an equivalence-key for 'fix all' purposes.

**Comment by mavasani** [comment]
_Note: from VS's perspective, a top level item can be an action itself. Though we currently don't expose any way to accomplish that with our CodeAction API._

When we add that support, we probably need better API to distinguish between:

1. A non-invocable aggregate top level action, whose equivalence key defaults to a value based on nested actions and

2. An invocable top level action, with flavored nested actions, whose equivalence key is independent of nested actions.

**Comment by Pilchie** [comment]
Fixed by #6297
